### PR TITLE
Migrate Commands: ('SETBIT', 'GETBIT', 'BITCOUNT', 'BITPOS', 'BITFIELD', 'BITFIELD_RO')

### LINKED ISSUE #1017: Command Migration: ('SETBIT', 'GETBIT', 'BITCOUNT', 'BITPOS', 'BITFIELD')
This issue tracks the migration of the mentioned commands - ('SETBIT', 'GETBIT', 'BITCOUNT', 'BITPOS', 'BITFIELD') to make them compatible across the three protocols supported by the Dice database: RESP, HTTP, and WebSocket.
DiceDB now supports more than one protocols (Resp/http/websocket). We don't want eval operation of each command to be strictly bound with any of the protocols. Currently eval function return statements of each command is bounded to the RESP protocol. 

*The migration is required because:*

- The current implementation is specific to the RESP protocol
- Evaluation functions of each command must be refactored to ensure protocol-agnostic behavior.
- Evaluation functions should return raw responses instead of protocol specific responses.

The goal is to make the command logic protocol-independent, allowing all three protocols to call the same core functionality seamlessly.

### Requirements

- Refactor the evaluation function of `[Command Name]` to be generic.
- Ensure the evaluation function does not include protocol-specific logic.

### Migration Steps

1. **Analyze Current Implementation**
   - Review the current codebase to understand how the command logic is implemented.

2. **Refactor Return Logic**
   - Create a new function with the function definition as `evalXXX(args []string, store *dstore.Store) *EvalResponse` under the file `/internal/eval/store_eval.go`
   - Analyse return statements of the eval and modify them to send raw types without encoding
   - Use errors in the return statements from `/errors/migrated_errors.go` file
   - Use perdefined responses from `/internal/clientio/resp.go` file as RespType variables

3. **Command/Worker specific Changes**
   - Make `IsMigrated` flag to true in the commands information under `/internal/eval/commands.go` file.
   - Use the newly written eval function against the `NewEval` parameter in the same command structure.
   - Delete the old eval function from the `/internal/eval/eval.go` file.
   - Add the migrated command to the `/internal/worker/CommandsMeta` map and make it's type as `SingleShard`

4. **Update Unit Tests**
   - Refactor existing unit tests to accommodate the new implementation.
   - Add new unit tests if necessary to cover all possible cases.

5. **Integration Tests**
   - Run all integration tests to ensure successful migration.
   - Ensure that each protocol (RESP, HTTP, WebSocket) works correctly after migration.

### Checklist

- [ ] Migrated the evalXXX function with the latest definition
- [ ] Update or add unit tests for the new implementation.
- [ ] All unit tests pass successfully.
- [ ] Ensure all integration tests pass successfully.

### Additional Notes

- Describe any edge cases that need to be handled in the generic function.
- Mention any protocol-specific optimizations required in wrappers.
  
If there are any questions or concerns about this migration, please mention them here.

### Related Issues/PRs

- Sample implementation for the Get, Set, GetSet and SetEx can be found in [this Pull Request.](https://github.com/DiceDB/dice/pull/766)

### PR DESCRIPTION
**Pull request by vishnuchandrashekar**
Migrates the eval functions for SETBIT, GETBIT, BITPOS, BITFIELD and BITFIELD_RO to the new eval function type independent of protocol.
Closes #1017 

Checklist:
- [x] Migrated the evalXXX function with the latest definition
- [x] Update or add unit tests for the new implementation. 
- [x] All unit tests pass successfully.
- [x] Ensure all integration tests pass successfully.
- [x] Ensure integration tests are added for the migrated command on multi-threaded resp server.
- [x] Move Integration tests for the respective commands under the RESP integration tests directory from Async directory
- [x] Please validate that the documentation for the respective commands is up to date. If not then consider adding them. 

Notes:
Since the BITFIELD and BITFIELD_RO commands depend on the same function (bitfieldEvalGeneric) I have gone ahead and migrated both the commands.

### CODE REVIEW

File: internal/errors/migrated_errors.go [hunk 1]
```diff
@@ -13,16 +13,19 @@ import (
 
 // Standard error variables for various DiceDB-related error conditions.
 var (
```
Comment (AshwinKul28) [inline internal/errors/migrated_errors.go:15]: Hi @vishnuchandrashekar any idea what has changed in this file. Need not to change unless adding any new error type.

File: internal/eval/store_eval.go [hunk 2]
```diff
@@ -316,3 +317,628 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  diceerrors.ErrWrongArgumentCount("SETBIT"),
+		}
+	}
+
+	key := args[0]
+	offset, err := strconv.ParseInt(args[1], 10, 64)
+	if err != nil {
+		return &EvalResponse{
+			Result: nil,
+			Error:  diceerrors.ErrIntegerOutOfRange,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:337]: Could have used `diceerrors.ErrGeneral` with the exact statement expected: `bit offset is not an integer or out of range`

File: internal/eval/store_eval.go [hunk 3]
```diff
@@ -316,3 +317,628 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+		}
+	}
+
+	value, err := strconv.ParseBool(args[2])
+	if err != nil {
+		return &EvalResponse{
+			Result: nil,
+			Error:  diceerrors.ErrIntegerOutOfRange,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:345]: Same as above, you can use `diceerrors.ErrGeneral` with the exact error message.

File: internal/eval/store_eval.go [hunk 4]
```diff
@@ -316,3 +317,628 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+		}
+		// newObj has bydefault expiry time -1 , we need to set it
+		if exDurationMs > 0 {
+			store.SetExpiry(newObj, exDurationMs)
+		}
+
+		store.Put(key, newObj)
+		if resp {
+			return &EvalResponse{
+				Result: 1,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:416]: Please use `clientio.IntegerOne`

File: internal/eval/store_eval.go [hunk 5]
```diff
@@ -316,3 +317,628 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+				Error:  nil,
+			}
+		}
+		return &EvalResponse{
+			Result: 0,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:421]: Please use `clientio.IntegerZero` same as other functions

**Review by AshwinKul28 — CHANGES_REQUESTED** [review]
Huge changes. Thanks a lot for this @vishnuchandrashekar . Left few minor comments please check and also check CI runs.

File: internal/errors/migrated_errors.go (hunk 1 above)
Comment (vishnuchandrashekar) [reply internal/errors/migrated_errors.go:15]: Hi, I've added the following error types: ErrInvalidBitfieldType, ErrBitfieldOffset, ErrOverflowType. Should I remove them and use `diceerrors.ErrGeneral`?

File: internal/eval/store_eval.go (hunk 4 above)
Comment (vishnuchandrashekar) [reply internal/eval/store_eval.go:416]: For some reason if I use `clientio.IntegerOne` and `clientio.IntegerZero` here the following tests fail: TestGetDel/Getdel_with_value_created_from_Setbit
TestType/TYPE_for_key_with_value_created_from_SETBIT_command.

I am not sure why.
Any suggestions?

File: internal/errors/migrated_errors.go (hunk 1 above)
Comment (AshwinKul28) [reply internal/errors/migrated_errors.go:15]: Ohh got it. Yes you can use General Errors since bit errors and even Overflow for a bit will not get used at anywhere else.

File: internal/eval/store_eval.go (hunk 4 above)
Comment (AshwinKul28) [reply internal/eval/store_eval.go:416]: This is failing in http tests. 
Please change the output of these tests in the following files

1. https://github.com/DiceDB/dice/blob/master/integration_tests/commands/http/type_test.go#L71 : float64(0) -> "0"
2. https://github.com/DiceDB/dice/blob/master/integration_tests/commands/http/getdel_test.go#L67 : float64(0) -> "0"

**Review by AshwinKul28 — APPROVED** [review]

**Comment by AshwinKul28** [comment]
HI @vishnuchandrashekar thanks a lot for the changes. Looks good. Please rebase your branch with master to resolve conflicts.

**Comment by soumya-codes** [comment]
@vishnuchandrashekar lets ensure that the corresponding integration tests are migrated as well.

**Comment by AshwinKul28** [comment]
HI @vishnuchandrashekar hope you are doing well. Do you still have any blockers on this PR? please let me or others know on the Discord.

**Comment by vishnuchandrashekar** [comment]
Hey @AshwinKul28, I've moved all the tests related to BIT to the file `integration_tests/commands/resp/bit_test.go`. Is this fine or do I need to do it the same way as it was done in `async`?

**Comment by apoorvyadav1111** [comment]
Hi @vishnuchandrashekar , you can rename the file. Additionally, we do not required the async test cases anymore, if you have migrated all the tests to resp, please remove the async tests. Additionally, please add tests in websocket and http folder. 

Thanks

**Comment by vishnuchandrashekar** [comment]
Hey @AshwinKul28 / @apoorvyadav1111, some of the websocket and http tests seems to be failing. I am unsure as to why. Any suggestions?

**Comment by apoorvyadav1111** [comment]
Hi @vishnuchandrashekar, I checked the lint run but there I can only see unit tests failing, Could you please share logs of failing websocket and HTTP tests on discord and tag me.
Thanks

**Comment by vishnuchandrashekar** [comment]
Hi @apoorvyadav1111, apologies I meant the tests that I added. Mainly the http BITPOS tests, could you have a look at those? I'll share the logs on discord soon. Thanks.

**Comment by apoorvyadav1111** [comment]
Yes, please share your logs on discord. Thanks

**Comment by JyotinderSingh** [comment]
Hi @vishnuchandrashekar are you still blocked on this issue?

**Comment by vishnuchandrashekar** [comment]
> Hi @vishnuchandrashekar are you still blocked on this issue?

Hi @JyotinderSingh, yes I'm still facing issues. I've sent a message on discord. Currently working on updating/adding the documentation for the commands.

**Comment by AshwinKul28** [comment]
HI @vishnuchandrashekar Changes look great. Please have a look at [this reference doc](https://github.com/DiceDB/dice/blob/master/docs/sample_command_docs.md) for the documentation changes. I see a few tests are failing, consider fixing them as well. Again, thanks a lot for the commendable efforts

**Comment by AshwinKul28** [comment]
HI @vishnuchandrashekar hope you are doing great. Can you please have a look at the documentation part as mentioned above? Also, is it possible to close this as early as possible? since we need to make DiceDB production ready soon for the upcoming major version release. 

Also more delay causes more conflicts 😭 to resolve as well. Let me know if you need any sort of help. Thanks again!

**Comment by vishnuchandrashekar** [comment]
Hey @AshwinKul28, I'll make sure to push the changes by the end of the day. Thanks for your patience.

**Comment by vishnuchandrashekar** [comment]
Hey @AshwinKul28, some of the integration tests that I've added are failing namely:

HTTP:
1. TestBitOpsString/Bitop_not_of_a_key_containing_a_string
2. TestBitOpsString/Bitop_not_of_a_key_containing_an_integer
3.  TestBitOpsString/BITOP_OR_of_keys_containing_strings_and_a_bytearray_and_get_the_destkey

RESP:
1. TestBitOpsString/BITOP_XOR_of_keys_containing_strings_and_get_the_destkey
2.  TestBitOpsString/BITOP_XOR_of_keys_containing_strings_and_a_bytearray_and_get_the_destkey

Websocket:
1. TestBitOpsString/Bitop_not_of_a_key_containing_a_string
2. TestBitOpsString/Bitop_not_of_a_key_containing_an_integer
3. TestBitOpsString/BITOP_OR_of_keys_containing_strings_and_a_bytearray_and_get_the_destkey

Also I am unsure as to why the build tests are failing, the tests pass on my local.
I am currently looking into this, would like your suggestions.

**Comment by vishnuchandrashekar** [comment]
Hey @AshwinKul28, I've fixed some of the failing test cases. I am still unsure why the RESP tests fail, would like your inputs.
@lucifercr07 - The http and websocket had some marshalling issues while returning the data, I've come up with a quick work around. Let me know what you think. I'm sure there is a better way to fix this.

**Comment by apoorvyadav1111** [comment]
Hi @vishnuchandrashekar , #1245 is closed. Lets move ahead.

**Review by apoorvyadav1111 — APPROVED** [review]
.....faaantastic work! Thanks for such a great effort.
